### PR TITLE
Filter/Projection Pushdown to Arrow Scans

### PR DESCRIPTION
**Pull request by pdet**
This PR implements Filter and Projection Pushdown directly to arrow scans (For both Python/R APIs).

In addition, this PR also implements a fetch arrow function in R (similar to the one that exists in Python) to transform DuckDB query results into Arrow Table 
```R
 arrow_table <-duckdb::duckdb_fetch_arrow(dbSendQuery(con, "SELECT * FROM test", arrow=TRUE))) 
```

cc @jonkeane @nealrichardson

### CODE REVIEW

**Comment by jonkeane** [comment]
Very cool on both fronts. I think I remember you mentioning that it might be possible to emit record batches instead of a table?

**Comment by jonkeane** [comment]
Additionally, while I was working on the wiring in Arrow[1], I did a comparison of with/without pushdown (since on the R side at least we can create the filters and then export the recordbatch reader with the filter, and the speed up was considerable (times + details are noted in the description of the PR): with pushdown the scan there took 27 seconds, whereas without it took 72 sec for the exact same aggregation calculations.

[1] https://github.com/apache/arrow/pull/10780

**Comment by pdet** [comment]
@jonkeane So that got merged in the arrow 5.0.0? Really cool!

Yes, I'll have a look into how to stream the results probably this week!

Really excited to see all this taking shape

**Comment by jonkeane** [comment]
It didn't quite make it in time for 5.0.0, our release candidate was cut a few days ago before this merged in — but it will be in 6.0.0. The R integration also relies on the upcoming release of DuckDB to be fully functional, IIRC that will be shortly, yeah?

**Comment by pdet** [comment]
Yes, (@hannesmuehleisen, please correct me if I'm wrong), but I think the plan is to make a new release either this or the next week.

File: tools/rpkg/R/register.R [hunk 1]
```diff
@@ -52,11 +52,12 @@ duckdb_register_arrow <- function(conn, name, arrow_scannable) {
   stopifnot(dbIsValid(conn))
 
     # create some R functions to pass to c-land
-    export_fun <- function(arrow_scannable, stream_ptr) {
-        record_batch_reader <- arrow::Scanner$create(arrow_scannable)$ToRecordBatchReader()
-        record_batch_reader$export_to_c(stream_ptr)
+    export_fun <- function(arrow_scannable, stream_ptr, projection=NULL, filter=TRUE) {
+        arrow::Scanner$create(arrow_scannable, projection, filter)$ToRecordBatchReader()$export_to_c(stream_ptr)
```
Comment (nealrichardson) [inline tools/rpkg/R/register.R:56]: FYI, if `arrow_scannable` is the result of previous dplyr queries on an Arrow object, `Scanner$create()` will [use the filter and projection stored in that object](https://github.com/apache/arrow/blob/master/r/R/dataset-scan.R#L82-L97) and will ignore your filter and projection args here

File: tools/rpkg/src/register.cpp [hunk 2]
```diff
@@ -58,25 +63,149 @@ class RArrowTabularStreamFactory {
...
+			    Rf_lang5(export_fun, factory->arrow_scannable, stream_ptr_sexp, projection_sexp, filters_sexp));
 		}
+		RApi::REvalThrows(export_call);
 		return res;
 	}
 
 	SEXP arrow_scannable;
 	SEXP export_fun;
+
+private:
```
Comment (nealrichardson) [inline tools/rpkg/src/register.cpp:97]: If I read this correctly, it looks like you're reimplementing how to build Arrow `Expression` objects to pass as a filter. This (and many many more functions) is all handled in the `arrow` package already--can you call the same code that our `dplyr::filter()` method does instead?

**Review by Mytherin — APPROVED** [review]

**Comment by Mytherin** [comment]
@pdet this is ready to merge now right?

**Comment by pdet** [comment]
Yeah, the tests all pass, neal's first comment will be resolved in arrow and the filter transformation we have is different from R->Arrow.

**Pull request merged**